Refactor gpu trace to be device-agnostic

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121795
* __->__ #121794

# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

guangyey commented on torch/_utils.py:
```diff
@@ -935,3 +938,25 @@ def queue_seed(self, cb, traceback):
+P = ParamSpec("P")
+
+
+class CallbackRegistry(Generic[P]):
```
Move `CallbackRegistry` from `torch.utils._cuda_trace.py` to here, which can be shared by each backend.

guangyey commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -18,17 +18,20 @@ namespace {
+    pybind11::gil_scoped_acquire gil;                                       \
+    try {                                                                   \
+      std::string module_name =                                             \
+          "torch." + at::DeviceTypeName(device_type, true) + "._gpu_trace"; \
```
An assumption: the callback functions will be located in `torch.xxx._gpu_trace.py` for each backend.

albanD replied:
You can use https://github.com/pytorch/pytorch/blob/be0bdf111ccfb403637e40dc26199c116b18cf73/torch/_utils.py#L891 here to simplify the logic

guangyey replied:
Do you mean like this,
```cpp
try {                                                                    \
  py::module utils_mod = py::module::import("torch._utils");             \
  py::object get_device_module = utils_mod.attr("_get_device_module");   \
  py::object hook = get_device_module(DeviceTypeName(device_type, true)) \
                          .attr("_gpu_trace")                            \
                          .attr(func_name)                               \
                          .attr("fire_callbacks");                       \
    hook(__VA_ARGS__);                                                   \
  } catch (const std::exception& e) {                                    \
    LOG(ERROR) << device_type                                            \
               << " trace hook execution failed: " << e.what();          \
  }                                                                      \
```

guangyey replied:
Option 2:
```cpp
try {                                                                       \
    std::string module_name = "torch." + DeviceTypeName(device_type, true); \
    py::module mod = py::module::import(module_name.c_str());               \
    py::object hook =                                                       \
        mod.attr("_gpu_trace").attr(func_name).attr("fire_callbacks");      \
    hook(__VA_ARGS__);                                                      \
  } catch (const std::exception& e) {                                       \
    LOG(ERROR) << device_type                                               \
               << " trace hook execution failed: " << e.what();             \
  }                                                                         \
```
option 2 is more concise, and Pybind can also raise a friendly error.
Which one do you prefer?

guangyey committed:
refactor gpu trace to device-agnostic

[ghstack-poisoned]

guangyey committed:
Update on "[WIP]refactor gpu trace to device-agnostic"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@albanD could you help review this PR?

EikanWang commented on test/test_cuda_trace.py:
```diff
@@ -19,18 +19,18 @@
 
     def test_event_creation_callback(self):
-        cuda_trace.register_callback_for_cuda_event_creation(self.mock)
+        gpu_trace.register_callback_for_event_creation(self.mock)
```
Since this test script is cuda-only. So you can set `cuda_trace` to `gpu_trace` like `cuda_trace = gpu_trace` to save loc.

guangyey commented on aten/src/ATen/cuda/CUDAEvent.h:
```diff
@@ -48,7 +48,7 @@ struct TORCH_CUDA_CPP_API CUDAEvent {
         const c10::impl::PyInterpreter* interp = c10::impl::GPUTrace::get_trace();
         if (C10_UNLIKELY(interp)) {
-          (*interp)->trace_gpu_event_deletion(reinterpret_cast<uintptr_t>(event_));
+          (*interp)->trace_gpu_event_deletion(at::kCUDA, reinterpret_cast<uintptr_t>(event_));
```
It is simpler to keep `PyInterpreter` as a singleton,

EikanWang replied:
How to correlate `singleton` and `at::kCUDA` as a parameter of `get_trace`? To take a parameter should not impact `singleton`

guangyey replied:
Sounds wired if `PyInterpreterVTable` has a private member related to device type.

albanD reviewed (approved):
Small change to simplify the c++ side, sounds good otherwise.

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Reverting to see if rocm trunk errors are related" -c weird

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@pytorchmergebot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it breaks ROCm jobs in trunk https://hud.pytorch.org/pytorch/pytorch/commit/74deacbf31d032a2659dc1633dc3e5248921d466, please help take a look and reland the change' -c nosignal

I'll add `ciflow/rocm` to run ROCm tests on the PR

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it breaks ROCm jobs in trunk https://hud.pytorch.org/pytorch/pytorch/commit/74deacbf31d032a2659dc1633dc3e5248921d466, please help take a look and reland the change' -c nosignal

guangyey commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -18,17 +18,24 @@ namespace {
+      /* Masquerade hip as cuda because hip uses `torch.cuda` module. */      \
+      if (device_type == at::kHIP) {                                          \
+        device_type = at::kCUDA;                                              \
+      }                                                                       \
```
@albanD In my opinion here needs your review again. The code change is that we need to specifically handle `hip` because `hip` uses `torch.cuda` module instead of `torch.hip`.

guangyey replied:
All CI failures seem unrelated.

albanD replied:
Ho that is pretty sad... Thanks for the update!

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "[Don't merge] Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "[Don't merge] Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@albanD May I know if you could help review this PR again>

albanD reviewed (approved):
Looks good!
Sorry for the delay, @huydhn feel free to ping me on these revert if I miss them!

guangyey commented:
> Looks good! Sorry for the delay, @huydhn feel free to ping me on these revert if I miss them!

Thank you very much~

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor gpu trace to be device-agnostic"


# Motivation
Refactor gpu trace to be device-agnostic. gpu trace is usually used in runtime components, including Device, Stream, Event, Guard, and Allocator. It should be device-agnostic and can be shared among each device backend.

# Solution
move `_cuda_trace.py` to `_gpu_trace.py`, which makes each device backend owns their callback, respectively.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge
